src/core/thread/mle_router.cpp

```diff
@@ -821,20 +819,44 @@ otError MleRouter::HandleLinkAccept(const Message &         aMessage,
         LogMleMessage("Receive Link Accept", aMessageInfo.GetPeerAddr(), sourceAddress.GetRloc16());
     }
 
-    // Version
-    SuccessOrExit(error = Tlv::GetTlv(aMessage, Tlv::kVersion, sizeof(version), version));
-    VerifyOrExit(version.IsValid(), error = OT_ERROR_PARSE);
+    VerifyOrExit(IsActiveRouter(sourceAddress.GetRloc16()), error = OT_ERROR_PARSE);
+
+    routerId      = GetRouterId(sourceAddress.GetRloc16());
+    router        = mRouterTable.GetRouter(routerId);
+    neighborState = (router != NULL) ? router->GetState() : Neighbor::kStateInvalid;
 
     // Response
     SuccessOrExit(error = Tlv::GetTlv(aMessage, Tlv::kResponse, sizeof(response), response));
     VerifyOrExit(response.IsValid(), error = OT_ERROR_PARSE);
 
+    // verify response
+    switch (neighborState)
+    {
+    case Neighbor::kStateLinkRequest:
+        VerifyOrExit(memcmp(router->GetChallenge(), response.GetResponse(), router->GetChallengeSize()) == 0,
+                     error = OT_ERROR_SECURITY);
+        break;
+
+    case Neighbor::kStateInvalid:
+    case Neighbor::kStateValid:
+        VerifyOrExit((mChallengeTimeout > 0) && (memcmp(mChallenge, response.GetResponse(), sizeof(mChallenge)) == 0),
+                     error = OT_ERROR_SECURITY);
+        break;
```

Comment:
Double-checking a change in https://github.com/openthread/openthread/pull/4297/commits/a656edb9c1d5bbbf0f371612758ac9786bb2884c:
- In the old code when `neighborState` was `kStateVlaid` we would not check the response.
- In new code we check it against `mChallenge`.

@jwhui , is this intentional/correct?

We seem to populate `mChallenge ` from `SendLinkRequest()` only when we have no neighbor.
 I think we can send a `Link Request` when the router/neighbor is already in valid state (e.g., when we miss "Link Avd" from router). In that case we seem to use a stack variable for challenge (not saved it in `mChallenge`):
https://github.com/openthread/openthread/blob/b69f463b3f69540fa4913e10cca715716da6ddf3/src/core/thread/mle_router.cpp#L516-L544